### PR TITLE
mplayer like speed controls

### PR DESCRIPTION
**Pull request by lucasw**
Press [ to slow playback speed 10%, or ] to increase speed 10%, or backspace to reset to rate 1.0  (though maybe if the user supplied a rate at the command line backspace should reset to that?).  

I'd like to add the ability to skip forward or back five seconds next, though my first attempt at using time_publisher_.setTime() hasn't worked.

### CODE REVIEW

**Comment by lucasw** [comment]
I figured out how to skip forward and backward any number of seconds, I'll put it in a future pull request.

**Comment by wjwwood** [comment]
Thanks for the pr @lucasw.

I'll have a look at the pr, but it might take a few days for all of us to review it, as some people are traveling.

File: tools/rosbag/src/player.cpp [hunk 1]
```diff
@@ -201,7 +202,7 @@ void Player::publish() {
     options_.advertise_sleep.sleep();
     std::cout << " done." << std::endl;
 
-    std::cout << std::endl << "Hit space to toggle paused, or 's' to step." << std::endl;
+    std::cout << std::endl << "Hit space to toggle paused, or 's' to step, [ or ] to decrease/increase playback speed, backspace to set speed to normal." << std::endl;
```
Comment (wjwwood) [inline tools/rosbag/src/player.cpp]: `Hit space to toggle paused, 's' to step, '['/']' to decrease/increase playback speed, or backspace to set speed to normal.`

File: tools/rosbag/src/player.cpp [hunk 2]
```diff
@@ -209,17 +210,25 @@ void Player::publish() {
         // Set up our time_translator and publishers
 
         time_translator_.setTimeScale(options_.time_scale);
-
+
```
Comment (wjwwood) [inline tools/rosbag/src/player.cpp]: Please remove any trailing whitespace.

File: tools/rosbag/src/player.cpp [hunk 3]
```diff
@@ -227,15 +236,20 @@ void Player::publish() {
             time_publisher_.setPublishFrequency(-1.0);
 
         paused_time_ = now_wt;
-
-        // Call do-publish for each message
```
Comment (wjwwood) [inline tools/rosbag/src/player.cpp:237]: Why remove this comment?

File: tools/rosbag/src/player.cpp [hunk 4]
```diff
@@ -364,6 +382,33 @@ void Player::doPublish(MessageInstance const& m) {
...
+                time_translator_.setTimeScale(options_.time_scale);
+                break;
+            case ',':
+                skip_time_ = true;
+                new_time_ = time_publisher_.getTime() - ros::Duration(5.0);
+                break;
+            case '.':
+                skip_time_ = true;
+                new_time_ = time_publisher_.getTime() + ros::Duration(5.0);
+                break;
```
Comment (wjwwood) [inline tools/rosbag/src/player.cpp]: This feature isn't mentioned in the line above that describes the possible hotkeys.

File: tools/rosbag/src/player.cpp [hunk 5]
```diff
@@ -329,7 +343,11 @@ void Player::doPublish(MessageInstance const& m) {
         bool charsleftorpaused = true;
         while (charsleftorpaused && node_handle_.ok())
         {
-            switch (readCharFromStdin()){
+            const int key = readCharFromStdin();
+            //if (key >= 0)
+            //    ROS_INFO_STREAM("key " << key);
```
Comment (wjwwood) [inline tools/rosbag/src/player.cpp]: You should change this back to just `switch(readCharFromStdin()) {`.

File: tools/rosbag/src/player.cpp [hunk 6]
```diff
@@ -227,15 +236,20 @@ void Player::publish() {
...
-        // Call do-publish for each message
+        
         foreach(MessageInstance m, view) {
             if (!node_handle_.ok())
                 break;
 
             doPublish(m);
+
+            if (skip_time_)
+                break;
```
Comment (wjwwood) [inline tools/rosbag/src/player.cpp]: Why wouldn't this either wrap around the `foreach` as an `if` statement or at least `break` before the first `doPublish`? Put another way, why does this only do one `doPublish` when `skip_time_` is `true`?

**Comment by wjwwood** [comment]
Ok, I've added some comments. I'll have to try out the pr later; the skipping seems tricky and seems like it might not cover all the corner cases, like skipping backwards at the beginning of a bag and skipping forwards at the end. It might handle these as is, but that's not clear to me from looking at the code.

File: tools/rosbag/src/player.cpp (hunk 4 above)
Comment (lucasw) [inline tools/rosbag/src/player.cpp]: I meant those changes for a future pull request (and to debug them more), but as it is I'll address the other comments and add the skip forward/back to the hotkey description.

File: tools/rosbag/src/player.cpp (hunk 6 above)
Comment (lucasw) [inline tools/rosbag/src/player.cpp]: `skip_time_` ought to always be false entering into the `foreach`, and is only set false within `doPublish` if a skip key is pressed.  (Though there are one or more unwanted doPublishes whenever I use the skip function, I see a glitch in rqt_plot when playing a bag with a steadily increasing published value and skipping forward or backward)

**Comment by lucasw** [comment]
Cool, thanks for all the comments!  The skipping definitely deserves more attention, there is logic there to set the time to `start_time_` when it is set earlier than it.  When skipping too far forward you should see it does exit rosbag if looping is off, and if looping is true it does loop, but I haven't tested the interactions with the other options yet.

**Comment by dirk-thomas** [comment]
Thanks for contributing new features like this.

As you mentioned before the `reset` operation should set the speed back to the initial rate value (which might be different from `1.0`).

Please also address the problem you mentioned regarding not exiting when fast forwarding and verify the interaction with other options.

**Comment by lucasw** [comment]
I'd like to put this onto https://github.com/lucasw/ros_comm/tree/mplayer_like_speed_controls (so my indigo-devel can track the one here)- is it possible to edit this pr to reflect that?  If not and I get back to this should I start a new pr anyhow?

**Comment by dirk-thomas** [comment]
You can't change the branch name used for a PR. In that case you need to open a new PR.

**Comment by lucasw** [comment]
Okay, closing this since it no longer points at the right branch, but I'll refer back to this upon making a new pr.

**Pull request closed without merging**